scripts/subspace_learning_lib.py

```diff
@@ -0,0 +1,243 @@
+    return projection_matrix
+
+
+@jit
+def adam_update(grads, params, mass, velocity, hyperparams):
+    mass = hyperparams["beta_1"] * mass + (1.0 - hyperparams["beta_1"]) * grads
+    velocity = hyperparams["beta_2"] * velocity + (1.0 - hyperparams["beta_2"]) * (grads ** 2.0)
+    # Bias correction
+    hat_mass = mass / (1 - hyperparams["beta_1"])
+    hat_velocity = velocity / (1 - hyperparams["beta_2"])
+    # Update
+    params = params - hyperparams["lr"] / (jnp.sqrt(hat_velocity) + hyperparams["epsilon"]) * hat_mass
+    return params, mass, velocity
+
+
+def loglikelihood(params, x, y, predict_fn):
+    """Computes the log-likelihood."""
+    logits = predict_fn(params, x)
+    ll = jnp.sum(logits[y])
+    return ll
+
+
+@jit
+def logprior(params_full, prior_variance=1.):
+    """Computes the Gaussian prior log-density."""
+    exp_term = -params_full ** 2 / (2 * prior_variance)
+    norm_constant = 0.5 * jnp.log((2 * jnp.pi * prior_variance))
+    return (exp_term - norm_constant).mean()
+
+
+def accuracy(params, batch, predict_fn):
+    inputs, targets = batch
+    logits = predict_fn(params, inputs)
+    predicted_class = jnp.argmax(logits, axis=-1)
+    return jnp.mean(predicted_class == targets), jnp.exp(logits)
+
+
+def data_stream(dataset, batch_size):
+    X, y = dataset
+    num_train = len(X)
+    num_batches, rem = divmod(num_train, batch_size)
+    num_batches += 1 if rem else 0
+    rng = np.random.RandomState(0)
+    while True:
+        perm = rng.permutation(num_train)
+        for i in range(num_batches):
+            batch_idx = perm[i * batch_size:min((i + 1) * batch_size, num_train)]
+            yield X[batch_idx], y[batch_idx]
+
+
+def inference_loop(rng_key, kernel, initial_state, num_samples):
+    @progress_bar_scan(num_samples)
+    def one_step(carry, i):
+        state, key = carry
+        kernel_key, key = split(key)
+        state, _ = kernel(kernel_key, state)
+        return (state, key), state
+
+    _, states = scan(one_step, (initial_state, rng_key), jnp.arange(num_samples))
+    return states
+
+
+def train_step(carry, i, dataloader, loss, hyperparams, callback):
+    batch = next(dataloader)
+    params, mass, velocity = carry
+    epoch_loss, grads = value_and_grad(loss)(params, batch)
+    params, mass, velocity = adam_update(grads, params, mass, velocity, hyperparams)
+    epoch_accuracy = callback(params, batch)
+    return (params, mass, velocity), (epoch_loss, epoch_accuracy)
+
+
+def build_nuts_sampler(num_warmup, potential):
+    def nuts_sampler(rng_key, num_samples, initial_params):
+        initial_state = nuts.new_state(initial_params, potential)
+
+        kernel_generator = lambda step_size, inverse_mass_matrix: jit(nuts.kernel(
+            potential, step_size, inverse_mass_matrix))
+
+        stan_key, key = split(rng_key)
+
+        final_state, (step_size, inverse_mass_matrix), _ = stan_warmup.run(
```

Comment:
if num_warmup = 0, you should have defaiults for `inverse_mass_matrix = jnp.eye()` and `step_size = 1e-3` or whatever.